File: packages/shared/src/contact-import.test.ts
/**
 * #248 round 3 — the per-column declaration that replaced the classifier.
 *
 * Two rounds of "does this dropped column mean do-not-contact" lost, once to a
 * vocabulary of words and once to a vocabulary of numbers. These are the tests
 * for the thing that has no vocabulary: every column is answered, by index.
 */
import { describe, expect, it } from "vitest";

import {
  CONTACT_IMPORT_COLUMN_FIELD,
  CONTACT_IMPORT_IGNORE,
  CONTACT_IMPORT_UNREADABLE_ENCODING,
  CONTACT_IMPORT_VCARD_PROPERTY_FIELD,
  contactImportColumnMismatchMessage,
  contactImportUndeclaredColumnsMessage,
  contactImportUndeclaredPropertiesMessage,
  contactImportUnterminatedQuoteMessage,
  defaultContactImportColumns,
  formatContactImportColumn,
  formatVCardProperty,
  mappingFromDeclarations,
  parseContactImportColumn,
  parseVCardProperty,
  vcardParameterProperty,
  VCARD_MAPPED_PROPERTIES,
} from "./contact-import";

describe("#248 the wire form of a column declaration", () => {
  it("round-trips a plain declaration", () => {
    const declaration = { index: 2, action: "opted_out" as const, header: "Do Not Call" };
    expect(formatContactImportColumn(declaration)).toBe("2:opted_out:Do Not Call");
    expect(parseContactImportColumn("2:opted_out:Do Not Call")).toEqual(declaration);
  });

  it("keeps a header containing colons, commas and quotes whole", () => {
    // The header is LAST in the string for exactly this reason: it may contain
    // anything a spreadsheet allows, and the two splits before it are on fixed
    // tokens. A format that split on every colon would truncate this header and
    // then report it as not matching the file.
    const raw = `0:${CONTACT_IMPORT_IGNORE}:Region: EU, "west"`;
    expect(parseContactImportColumn(raw)).toEqual({
      index: 0,
      action: CONTACT_IMPORT_IGNORE,
      header: `Region: EU, "west"`,
    });
  });

  it("carries a NAMELESS column, which is the whole reason it is index-first", () => {
    // Round two matched its field on `normalizeContactHeader`, which strips
    // everything but [a-z0-9] — so "", "—", "#" and "★" all normalised to the
    // SAME EMPTY STRING and two nameless columns could not be told apart. A
    // position cannot collide with another position.
    expect(parseContactImportColumn("3:ignore:")).toEqual({
      index: 3,
      action: CONTACT_IMPORT_IGNORE,
      header: "",
    });
    const first = parseContactImportColumn("2:ignore:—");
    const second = parseContactImportColumn("3:ignore:★");
    expect(first?.index).not.toBe(second?.index);
  });

  it("refuses anything that is not a declaration", () => {
    for (const raw of [
      "",
      "2",
      "2:opted_out",
      "two:ignore:x",
      " 2:ignore:x", // a lenient Number() would read this as column two
      "+2:ignore:x",
      // NO INDEX AT ALL. The one the strictness above missed, because
      // `Number("")` is 0: a single `rawIndex !== ""` escape hatch turns every
      // indexless declaration into a confident answer about column ZERO, which
      // is identity-by-header again under a new spelling — and header identity
      // is what could not tell two nameless columns apart.
      ":ignore:x",
      ":ignore:",
      "::x",
      "2:phone_number:x", // not a field this importer has
      "2::x",
      "-1:ignore:x",
    ]) {
      expect(parseContactImportColumn(raw)).toBeNull();
    }
  });
});

describe("#248 the default guess a client shows before anybody answers", () => {
  it("H1: guesses a FIELD or nothing, and never a dismissal", () => {
    // THE SHIP BLOCKER of round three, in one assertion. This used to answer
    // `ignore` for every column it did not recognise, so `Phone,Name,Notes`
    // over a Notes column reading "DO NOT CALL - asked us to stop" came back a
    // COMPLETE declaration: all three clients posted it with no interaction,
    // the API accepted it because complete is all it can check, and the send
    // returned 201 with a message row. A dismissal is an answer — "I saw these
    // values and they decide nothing" — and a detector has seen nothing.
    const headers = ["Phone Number", "Full Name", "Marketing Status"];
    expect(defaultContactImportColumns(headers)).toEqual([
      { index: 0, action: "phone", header: "Phone Number" },
      { index: 1, action: "name", header: "Full Name" },
      { index: 2, action: null, header: "Marketing Status" },
    ]);
    // And it is not merely "not this literal": nothing this function returns is
    // an action at all unless the detector recognised a field. Asserted against
    // the shipped constant, so a rename of `ignore` moves the guard with it.
    for (const guess of defaultContactImportColumns([
      "Notes",
      "Marketing Status",
      "Owner",
    ])) {
      expect(guess.action).not.toBe(CONTACT_IMPORT_IGNORE);
    }
  });

  it("still lists EVERY column, including one that exists only in the data", () => {
    // Unanswered is not absent: the client renders these by position, so a
    // column missing from this list is a column missing from the screen — and
    // a cell past the end of the header row is exactly the column somebody
    // annotated one row with.
    const guesses = defaultContactImportColumns(
      ["Phone", "Name"],
      [["+12065550101", "Ann", "DO NOT CALL"]],
    );
    expect(guesses).toHaveLength(3);
    expect(guesses[2]).toEqual({ index: 2, action: null, header: "" });
    // A MIX: the recognised columns keep their guess in the same answer, so
    // nobody retypes `Phone` → phone on a clean export. A gate that fires on
    // every file is a gate people learn to click through.
    expect(guesses.map((guess) => guess.action)).toEqual([
      "phone",
      "name",
      null,
    ]);
  });
});

describe("#248 the mapping the declarations describe", () => {
  it("is the person's answer, not the detector's", () => {
    // `Description` is claimed by `notes` on every header pattern we own, so a
    // row reading "DO NOT CONTACT" under it was filed as a note and the person
    // was texted. Round two's gate only examined UNMAPPED columns, so a wrong
    // mapping was invisible to it. Here the human's answer wins outright.
    expect(
      mappingFromDeclarations([
        { index: 0, action: "phone", header: "Phone" },
        { index: 1, action: "opted_out", header: "Description" },
        { index: 2, action: CONTACT_IMPORT_IGNORE, header: "Owner" },
      ]),
    ).toEqual({ phone: 0, opted_out: 1 });
  });

  it("drops the ignored columns entirely", () => {
    expect(
      mappingFromDeclarations([
        { index: 0, action: CONTACT_IMPORT_IGNORE, header: "a" },
      ]),
    ).toEqual({});
  });
});

describe("#248 what the refusals say", () => {
  it("names each undeclared column by position, header or not", () => {
    const message = contactImportUndeclaredColumnsMessage(
      [
        { index: 2, header: "Marketing Status" },
        { index: 3, header: "" },
      ],
      4,
    );
    expect(message).toContain(`column 3 ("Marketing Status")`);
    expect(message).toContain("column 4 (no header)");
    // The field name comes from the shipped constant, so a rename moves the
    // sentence with it rather than leaving a caller following stale advice.
    expect(message).toContain(`\`${CONTACT_IMPORT_COLUMN_FIELD}\``);
    expect(message).toContain("Nothing was imported.");
  });

  it("says the plural and the singular", () => {
    expect(
      contactImportUndeclaredColumnsMessage([{ index: 0, header: "x" }], 1),
    ).toContain("was not declared");
    expect(
      contactImportUndeclaredColumnsMessage(
        [
          { index: 0, header: "x" },
          { index: 1, header: "y" },
        ],
        2,
      ),
    ).toContain("were not declared");
    // The grammar every refused workspace reads. Round two shipped "this import
    // does not know what those columns means" to exactly these people.
    expect(
      contactImportUndeclaredColumnsMessage([{ index: 0, header: "x" }], 1),
    ).not.toMatch(/columns means|column mean\b/);
  });

  it("keeps the mismatch sentence one sentence, whatever went wrong", () => {
    const message = contactImportColumnMismatchMessage("column 2 is declared twice");
    expect(message).toContain("column 2 is declared twice");
    expect(message).toContain("Nothing was imported.");
  });
});

describe("#248 the vCard property declaration", () => {
  it("maps only the properties the importer actually reads", () => {
    // If `CATEGORIES` or `NOTE` ever appears here, the two places a .vcf can
    // say do-not-text stop being asked about.
    expect(VCARD_MAPPED_PROPERTIES).not.toContain("CATEGORIES");
    expect(VCARD_MAPPED_PROPERTIES).not.toContain("NOTE");
    expect([...VCARD_MAPPED_PROPERTIES].sort()).toEqual([
      "BEGIN",
      "END",
      "FN",
      "N",
      "TEL",
      "VERSION",
    ]);
  });

  it("round-trips, and upper-cases the property", () => {
    expect(formatVCardProperty({ property: "CATEGORIES", action: "opted_out" })).toBe(
      "CATEGORIES:opted_out",
    );
    expect(parseVCardProperty("categories:opted_out")).toEqual({
      property: "CATEGORIES",
      action: "opted_out",
    });
  });

  it("refuses anything that is not a property declaration", () => {
    for (const raw of ["", "NOTE", "NOTE:notes", ":ignore", "NOTE:IGNORE"]) {
      expect(parseVCardProperty(raw)).toBeNull();
    }
  });

  it("names the properties it will not read past", () => {
    const message = contactImportUndeclaredPropertiesMessage(["CATEGORIES", "NOTE"]);
    expect(message).toContain("`CATEGORIES`");
    expect(message).toContain("`NOTE`");
    expect(message).toContain(`\`${CONTACT_IMPORT_VCARD_PROPERTY_FIELD}\``);
    expect(message).toContain("Nothing was imported.");
  });

  it("H3: carries a PARAMETER as its own property, through the same wire form", () => {
    // `TEL;TYPE=CELL;X-ABLabel=DO NOT CALL:+1613…` is Apple's inline shape, and
    // the parameter is where the instruction sat. It is declared like any other
    // property, which works only because the wire form splits on the LAST colon
    // — a format that split on the first would read this as a property called
    // `TEL` and an action called `TYPE`.
    const token = vcardParameterProperty("TEL", "X-ABLABEL");
    expect(token).toBe("TEL;X-ABLABEL");
    expect(formatVCardProperty({ property: token, action: "opted_out" })).toBe(
      "TEL;X-ABLABEL:opted_out",
    );
    expect(parseVCardProperty("TEL;X-ABLABEL:opted_out")).toEqual({
      property: token,
      action: "opted_out",
    });
  });

  it("H3: says how many properties it did not print, rather than trailing off", () => {
    // A line with no colon is now a property, so a mangled file can carry
    // thousands of one-off tokens as long as the line. Printing them all is a
    // refusal nobody reads; printing some and saying nothing is #248 round 2's
    // B8 defect — a truncated list under a whole count, which people act on.
    const many = Array.from({ length: 25 }, (_, i) => `X-PROP-${i}`);
    const message = contactImportUndeclaredPropertiesMessage(many);
    expect(message).toContain("`X-PROP-0`");
    expect(message).toContain("and 5 more");
    expect(message).not.toContain("`X-PROP-24`");
    // One long token is shortened rather than reprinted whole, and says so.
    const long = `X-${"D".repeat(200)}`;
    const shortened = contactImportUndeclaredPropertiesMessage([long]);
    expect(shortened).not.toContain(long);
    expect(shortened).toContain("…");
    // A file with an ordinary number of unread properties prints all of them
    // and never says "more" — the branch has to have both sides.
    expect(
      contactImportUndeclaredPropertiesMessage(["CATEGORIES", "NOTE"]),
    ).not.toContain("more");
  });
});

describe("#248 H5 what the server says about a file it cannot read", () => {
  it("names the line a quote opened on, and says nothing was imported", () => {
    // The mangling shows up hundreds of lines later, where the file looks fine,
    // so the line the quote OPENED on is the only part a person can act on.
    const message = contactImportUnterminatedQuoteMessage(3);
    expect(message).toContain("line 3");
    expect(message).toContain("Nothing was imported.");
    // And it says what the silent version cost: rows that are neither imported
    // nor reported. A refusal that reads as "some rows were skipped" would send
    // somebody looking at the wrong thing.
    expect(message).toContain("would not be reported as skipped");
  });

  it("tells a UTF-16 save what to do instead of answering 500", () => {
    expect(CONTACT_IMPORT_UNREADABLE_ENCODING).toContain("UTF-16");
    expect(CONTACT_IMPORT_UNREADABLE_ENCODING).toContain("CSV UTF-8");
    expect(CONTACT_IMPORT_UNREADABLE_ENCODING).toContain("Nothing was imported.");
  });
});

File: apps/android/app/src/test/kotlin/com/loonext/android/core/contacts/ContactImportTest.kt
package com.loonext.android.core.contacts

import com.loonext.android.core.i18n.AppStrings
import java.io.File
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * #248 — the Kotlin port of `packages/shared/src/contact-import.ts`, checked
 * against the TypeScript itself rather than against numbers typed in here.
 *
 * A test that pins `MAX_ROWS == 2000` proves the port has not changed. It does
 * not prove the port is RIGHT, and the day the shared cap moves it becomes a
 * ceiling holding the port at last month's value — which is exactly how the
 * attestation field came to exist on the server and nowhere else. So the
 * source of truth is read from the repo, the same way `ParityVectorsTest`
 * reads its vectors.
 */
class ContactImportTest {

    /**
     * #228 — the ENGLISH behind a catalogue key.
     *
     * Half of this class asks what a sentence SAYS, and since #228 the copy
     * objects hold keys rather than sentences. Reading the key's English back
     * out of the catalogue keeps every one of those questions intact, and
     * throwing on a missing key means a sentence deleted from the catalogue
     * fails here rather than quietly rendering its own key name on a phone.
     */
    private fun en(key: String): String =
        AppStrings.en[key] ?: throw AssertionError("no English for $key")

    /**
     * Walk UP to the repo root: Gradle runs unit tests from `apps/android/app`,
     * but that is a property of the runner, not a promise.
     */
    private fun sharedSource(): String {
        var dir: File? = File("").absoluteFile
        while (dir != null) {
            val candidate = File(dir, "packages/shared/src/contact-import.ts")
            if (candidate.exists()) return candidate.readText()
            dir = dir.parentFile
        }
        throw AssertionError(
            "packages/shared/src/contact-import.ts not found walking up from " +
                File("").absolutePath,
        )
    }

    /** `export const NAME = "literal";` */
    private fun stringConst(source: String, name: String): String {
        val match = Regex("""const $name\s*=\s*\n?\s*"([^"]*)"""").find(source)
            ?: throw AssertionError("$name is not a string constant in the shared module")
        return match.groupValues[1]
    }

    /**
     * `export const NAME = 2000;` or `= 2 * 1024 * 1024;` or `= OTHER_CONST;`.
     * Products and aliases are both in the shared file, and resolving them here
     * beats asserting against a flattened number nobody can trace back.
     */
    private fun numberConst(source: String, name: String): Long {
        val match = Regex("""const $name\s*=\s*([A-Z0-9_ *]+);""").find(source)
            ?: throw AssertionError("$name is not a numeric constant in the shared module")
        return match.groupValues[1]
            .split("*")
            .map { it.trim() }
            .fold(1L) { product, term ->
                product * (term.toLongOrNull() ?: numberConst(source, term))
            }
    }

    @Test
    fun `the attestation field and value match the shared module`() {
        val source = sharedSource()
        // If either drifts, every import from this app 422s against a field
        // name the UI has no control for — which is precisely what #226 did.
        assertEquals(
            stringConst(source, "CONTACT_IMPORT_CONSENT_FIELD"),
            ContactImport.CONSENT_FIELD,
        )
        assertEquals(
            stringConst(source, "CONTACT_IMPORT_CONSENT_VALUE"),
            ContactImport.CONSENT_VALUE,
        )
    }

    @Test
    fun `the bounds this app prints match the shared module`() {
        val source = sharedSource()
        assertEquals(
            numberConst(source, "CONTACT_IMPORT_MAX_ROWS"),
            ContactImport.MAX_ROWS.toLong(),
        )
        assertEquals(numberConst(source, "CONTACT_IMPORT_MAX_BYTES"), ContactImport.MAX_BYTES)
        assertEquals(
            numberConst(source, "VCARD_IMPORT_MAX_CARDS"),
            ContactImport.VCARD_MAX_CARDS.toLong(),
        )
        assertEquals(numberConst(source, "VCARD_IMPORT_MAX_BYTES"), ContactImport.VCARD_MAX_BYTES)
    }

    @Test
    fun `each import kind carries the bounds the server enforces for it`() {
        assertEquals(ContactImport.MAX_ROWS, ContactImportKind.CSV.maxEntries)
        assertEquals(ContactImport.MAX_BYTES, ContactImportKind.CSV.maxBytes)
        assertEquals(ContactImport.VCARD_MAX_CARDS, ContactImportKind.VCARD.maxEntries)
        assertEquals(ContactImport.VCARD_MAX_BYTES, ContactImportKind.VCARD.maxBytes)
    }

    @Test
    fun `an unattested import posts no consent field at all`() {
        // Fail CLOSED, and fail on the SERVER: the app never decides locally
        // that consent is unnecessary, it simply declines to make the claim.
        assertEquals(
            emptyList<Pair<String, String>>(),
            ContactImport.csvFields(false, emptyList()),
        )
        assertEquals(
            emptyList<Pair<String, String>>(),
            ContactImport.vcardFields(false, emptyList()),
        )
    }

    @Test
    fun `an attested import posts exactly the field the server demands`() {
        assertEquals(
            listOf(ContactImport.CONSENT_FIELD to ContactImport.CONSENT_VALUE),
            ContactImport.csvFields(true, emptyList()),
        )
        // The server accepts the literal "true" and nothing else — a value that
        // could also be "false" is a field, not an attestation.
        assertEquals("true", ContactImport.CONSENT_VALUE)
    }

    @Test
    fun `the two declaration fields match the shared module`() {
        // #248 round 3. If either drifts, every declaration is posted under a
        // name the server does not read — and since the gate is unconditional,
        // that is not a degraded import, it is no import at all.
        val source = sharedSource()
        assertEquals(
            stringConst(source, "CONTACT_IMPORT_COLUMN_FIELD"),
            ContactImport.COLUMN_FIELD,
        )
        assertEquals(
            stringConst(source, "CONTACT_IMPORT_VCARD_PROPERTY_FIELD"),
            ContactImport.VCARD_PROPERTY_FIELD,
        )
    }

    @Test
    fun `round two's acknowledgement field is gone from both sides`() {
        // It was sent only for the columns the server had just complained about,
        // so the shortest path to a 200 was: post, read the column names out of
        // the 422, post again. Two round trips and no human — demonstrated live.
        // A complete declaration removes the loop rather than policing it.
        assertFalse(sharedSource().contains("REVIEWED_COLUMN"))
        assertFalse(
            readMainSource("core/contacts/ContactImport.kt").contains("REVIEWED_COLUMN_FIELD"),
        )
    }

    @Test
    fun `a declaration is posted once per column, under a repeated name`() {
        // The field is REPEATED — a header may contain anything, commas and
        // colons included, so there is no delimiter to get wrong — and a Map
        // cannot say a name twice. That is why these are pairs: the shape this
        // app posted in until #248 could not have expressed the answer at all.
        val columns = listOf("0:phone:Phone", "1:ignore:Region, EU", "2:opted_out:Do: Not Call")
        val fields = ContactImport.csvFields(true, columns)
        assertEquals(
            listOf(ContactImport.CONSENT_FIELD to ContactImport.CONSENT_VALUE) +
                columns.map { ContactImport.COLUMN_FIELD to it },
            fields,
        )
        assertEquals(
            "every column must survive, under one repeated field name",
            3,
            fields.count { it.first == ContactImport.COLUMN_FIELD },
        )
    }

    @Test
    fun `the vCard door posts its properties under its own field`() {
        // Two field names because they are two different questions. A `column`
        // posted at the vCard door would be read by nothing and the cards'
        // CATEGORIES would go undeclared.
        val fields = ContactImport.vcardFields(true, listOf("CATEGORIES:opted_out", "NOTE:ignore"))
        assertEquals(
            listOf(
                ContactImport.CONSENT_FIELD to ContactImport.CONSENT_VALUE,
                ContactImport.VCARD_PROPERTY_FIELD to "CATEGORIES:opted_out",
                ContactImport.VCARD_PROPERTY_FIELD to "NOTE:ignore",
            ),
            fields,
        )
        assertTrue(fields.none { it.first == ContactImport.COLUMN_FIELD })
    }

    @Test
    fun `neither declaration can be defaulted`() {
        // THE safety property of these two functions, and it is enforced by the
        // compiler rather than by this assertion — there is no overload to call
        // without an answer. The assertion is on the SOURCE because a default
        // added later would compile, pass every other test here, and put the
        // silent case back: a caller reaching the import route with nobody
        // having said what the file contains.
        val source = readMainSource("core/contacts/ContactImport.kt")
        for (signature in listOf("columns: List<String>", "properties: List<String>")) {
            assertTrue("$signature is gone", source.contains(signature))
            assertFalse(
                "$signature must not gain a default — see the docblock",
                source.contains("$signature = "),
            )
        }
        assertFalse(
            "and neither may `attested`, for the same reason",
            source.contains("attested: Boolean = "),
        )
    }

    /** This module's own source, for the assertions a value cannot express. */
    private fun readMainSource(relative: String): String {
        val bases = listOf(
            "src/main/kotlin/com/loonext/android",
            "app/src/main/kotlin/com/loonext/android",
            "apps/android/app/src/main/kotlin/com/loonext/android",
        )
        for (base in bases) {
            val file = File(File(base), relative)
            if (file.exists()) return file.readText()
        }
        throw AssertionError("source not found: $relative (cwd=${File(".").absolutePath})")
    }

    @Test
    fun `the overflow line counts what the server said, not what fitted`() {
        // #248 round 2 (B5/B8): the refusal count and the refusal list are two
        // facts, and until the count stopped always being zero they could never
        // disagree. A screen printing "40 refused" over five rows with nothing
        // saying the rest existed is the failure this closes, and every refused
        // row names a person this workspace must not text.
        assertEquals("…and 35 more.", ContactImport.overflowLine(total = 40, listed = 5))
        assertNull("a whole list needs no line", ContactImport.overflowLine(total = 5, listed = 5))
        // Never negative, never "…and -3 more." — a list longer than its count
        // is a server bug, and inventing an apology for it helps nobody.
        assertNull(ContactImport.overflowLine(total = 2, listed = 5))
        assertNull(ContactImport.overflowLine(total = 0, listed = 0))
    }

    @Test
    fun `a column's values are shown so it can be recognised`() {
        // THE reason this screen exists. A header alone is often not enough —
        // "Status" means nothing until you see it holds `active` and
        // `unsubscribed`, and at that point the answer is obvious and the
        // opposite of the one a skimmer would give.
        val line = ContactImport.Columns.valuesLine(listOf("active", "unsubscribed"))
        assertTrue(line, line.contains("active"))
        assertTrue(line, line.contains("unsubscribed"))
        // "include", never "holds": these are the first ImportColumns.SAMPLE_LIMIT
        // distinct values, and claiming to have listed the column would be a
        // claim nobody checked.
        assertTrue(line, line.contains("include"))
        assertFalse(line, line.lowercase().contains("holds"))
        // An empty column is still asked about, so it still needs a sentence —
        // and the sentence has to say it is empty rather than say nothing.
        assertTrue(
            ContactImport.Columns.valuesLine(emptyList()),
            ContactImport.Columns.valuesLine(emptyList()).contains("empty"),
        )
    }

    @Test
    fun `every answer a column may carry has a label of its own`() {
        // A control offering a raw `first_name` is a control nobody can read, and
        // two options sharing a label is a control nobody can use.
        val labels = ImportColumns.ACTIONS.map { ContactImport.Columns.actionLabel(it) }
        assertEquals("no two answers may read the same", labels.size, labels.toSet().size)
        for ((action, label) in ImportColumns.ACTIONS.zip(labels)) {
            assertTrue(action, label.isNotBlank())
            assertFalse("$action is printed raw", label == action)
        }
        // The two consequential ones say what they DO, in the words somebody
        // looking at a spreadsheet column would use.
        assertEquals("Do not text", ContactImport.Columns.actionLabel(ImportColumns.FIELD_OPTED_OUT))
        assertTrue(
            ContactImport.Columns.actionLabel(ImportColumns.ACTION_IGNORE)
                .lowercase()
                .contains("ignore"),
        )
    }

    @Test
    fun `the per-column step names the consequence of a wrong answer`() {
        // Somebody who ignores a do-not-call column here texts everybody it was
        // protecting, and that sentence has to be in front of them while they
        // answer — naming the outcome, not the procedure.
        // #228: built through the catalogue now, and the `{answer}` slot is
        // still filled from the control's own label rather than retyped.
        val warning = ContactImport.Columns.wrongColumn()
        assertTrue(warning, warning.contains(ContactImport.Columns.actionLabel("opted_out")))
        assertTrue(warning, warning.contains("texted"))
        assertTrue(warning, warning.contains("protecting"))
        // And the LEAD says why the question is being asked at all. "We did not
        // recognise these" invites "so skip them", which is the behaviour that
        // caused #248 twice.
        val lead = en(ContactImport.Columns.LEAD)
        assertTrue(lead, lead.contains("does not guess"))
    }

    @Test
    fun `a column is named by position as well as by header`() {
        // The server names a column at fault by position ("column 3"), and two
        // columns in one file are allowed to share a header — a name on its own
        // cannot always say which. 1-based, so the two agree.
        assertEquals("Column 3", ContactImport.Columns.positionLabel(2))
        assertEquals("Column 1", ContactImport.Columns.positionLabel(0))
        // A nameless column is asked about like any other. Round two withheld
        // these because its field matched columns BY NAME and no answer would
        // work; the declaration is by index, so there is now an answer — which is
        // what lets a cell past the end of the header row be accounted for.
        assertTrue(ContactImport.Columns.headerLabel("").contains("no header"))
        assertTrue(ContactImport.Columns.headerLabel("   ").contains("no header"))
        assertTrue(ContactImport.Columns.headerLabel("Do Not Call").contains("Do Not Call"))
    }

    @Test
    fun `the progress line counts what is answered, and does not start at zero`() {
        // It opens at whatever the detector recognised. A progress line reading
        // "0 of 7" on a screen that is four-sevenths finished is the reason
        // people abandon a flow — and it would also be a lie.
        assertEquals("4 of 7 answered", ContactImport.Columns.progressLine(4, 7))
        assertEquals("0 of 1 answered", ContactImport.Columns.progressLine(0, 1))
    }

    @Test
    fun `a field claimed twice is named before the server has to refuse it`() {
        // "a contact has one" — the server's own words. Catching it on the sheet
        // only ever SAVES a round trip; it cannot let anything through, which is
        // why a form check here is not a gate here.
        val hint = ContactImport.Columns.duplicateHint("phone")
        assertTrue(hint, hint.contains(ContactImport.Columns.actionLabel("phone")))
        assertTrue(hint, hint.contains("Two columns"))
    }

    @Test
    fun `the vCard sheet says how blunt its blocking answer is`() {
        // Declaring a property `opted_out` blocks EVERY card carrying it, whatever
        // it says — a CATEGORIES of "Friends" alongside one of "DNC". Genuinely
        // coarse, and somebody should know before they choose it rather than
        // after.
        val coarse = ContactImport.Properties.coarse()
        assertTrue(coarse, coarse.contains("every card"))
        assertTrue(
            coarse,
            coarse.contains(
                ContactImport.Properties.actionLabel(VCardProperties.ACTION_OPTED_OUT),
            ),
        )
        // And the lead says where a .vcf can say do-not-text at all, because
        // nobody knows that: it is a note or a category, and nothing else.
        val lead = en(ContactImport.Properties.LEAD)
        assertTrue(lead, lead.contains("note"))
        assertTrue(lead, lead.contains("categories"))
        val labels = VCardProperties.ACTIONS.map { ContactImport.Properties.actionLabel(it) }
        assertEquals(labels.size, labels.toSet().size)
        for ((action, label) in VCardProperties.ACTIONS.zip(labels)) {
            assertFalse("$action is printed raw", label == action)
        }
    }

    @Test
    fun `every printed figure is derived from the bounds, not typed`() {
        val csvMb = ContactImport.MAX_BYTES / (1024L * 1024L)
        val vcardMb = ContactImport.VCARD_MAX_BYTES / (1024L * 1024L)

        // The two doors have different caps, so the two sentences must differ —
        // a copy-paste that quoted the CSV cap at a vCard would promise a file
        // the server refuses, at the exact moment a new crew is least patient.
        assertTrue(
            ContactImport.tooLargeMessage(ContactImportKind.CSV).contains("$csvMb MB"),
        )
        assertTrue(
            ContactImport.tooLargeMessage(ContactImportKind.VCARD).contains("$vcardMb MB"),
        )
        assertTrue(
            ContactImport.limitsLine(ContactImportKind.CSV)
                .contains("${ContactImport.MAX_ROWS / 1000},000 rows"),
        )
        assertTrue(
            ContactImport.limitsLine(ContactImportKind.VCARD)
                .contains("${ContactImport.VCARD_MAX_CARDS / 1000},000 cards"),
        )
        assertTrue(
            ContactImport.limitsLine(ContactImportKind.VCARD).contains("$vcardMb MB"),
        )
    }

    @Test
    fun `neither door claims an import can lift a STOP`() {
        // The server can only ever ADD a block: the CSV path revives a REVOKED
        // opt-out and inserts new ones, the vCard path never touches opt_outs.
        // Copy that hinted otherwise would be the one promise about carrier
        // truth we are not allowed to get wrong.
        // #228: `optOutNote` is a catalogue key, so the sentence is read back
        // out of the English map. STOP is a carrier keyword and stays STOP in
        // both languages — a translated STOP is an opt-out nothing registers.
        for (kind in ContactImportKind.entries) {
            assertTrue(kind.name, en(kind.optOutNote).contains("STOP"))
            assertTrue(kind.name, AppStrings.frCA[kind.optOutNote]?.contains("STOP") == true)
        }
        assertTrue(en(ContactImportKind.CSV.optOutNote).contains("opted-out column"))
        // #248 round 3 changed the FACT this sentence reports, so the guard had
        // to change with it: a .vcf CAN now say do-not-text, through a property
        // somebody declared. The old wording ("a .vcf carries no opt-out column")
        // was true of the parser and false of the format, and that gap is how
        // CATEGORIES:DNC came to be dropped in silence.
        assertTrue(en(ContactImportKind.VCARD.optOutNote).contains("do-not-text"))
        for (kind in ContactImportKind.entries) {
            assertFalse(
                "${kind.name} must not claim a card cannot carry an opt-out",
                en(kind.optOutNote).contains("no opt-out"),
            )
        }
    }

    @Test
    fun `the refusal heading counts the rows it is about`() {
        // The server sends the number; the heading exists to say it. "Some
        // customers" over a list of forty is the count nobody can act on, and
        // this figure is the one a carrier audit actually asks for.
        assertTrue(ContactImport.consentRefusedHeadline(40).contains("40"))
        assertTrue(ContactImport.consentRefusedHeadline(1).contains("1"))
        // And it says what the count is OF — a bare number over a red panel
        // reads as a failure, which this is not.
        for (count in listOf(1, 2, 40)) {
            assertTrue(
                count.toString(),
                ContactImport.consentRefusedHeadline(count).contains("opted out"),
            )
        }
    }

    @Test
    fun `the refusal heading reads correctly for a single customer`() {
        // One refused row is the common case — a crew importing a book they
        // have texted before hits exactly one old STOP — so "1 customers" is
        // the version most people would see.
        assertFalse(
            ContactImport.consentRefusedHeadline(1),
            ContactImport.consentRefusedHeadline(1).contains("customers"),
        )
        assertTrue(ContactImport.consentRefusedHeadline(1).contains("customer"))
        assertTrue(ContactImport.consentRefusedHeadline(2).contains("customers"))
    }

    @Test
    fun `the sheet states what ticking the box actually records`() {
        // The server records the attestation ONLY where there is no basis yet
        // and leaves an existing one alone (importConsentColumns). Copy saying
        // "recorded for everyone" would describe the behaviour #248 removed.
        val recorded = en(ContactImport.Copy.RECORDED)
        assertTrue(recorded, recorded.contains("no consent recorded yet"))
        assertTrue(recorded, recorded.contains("keep it"))
    }
}
